Quinoa dev server forwards to ipv6 loopback address, instead of ipv4

Hello!

I'm trying to embedd a vite vue frontend into my quarkus app. I've followed the steps here: https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#vite

While runing `quarkus dev` it starts up without issues. But when attempting to reach localhost:8080, I get this issue where Quinoa attempts to forward requests to the ipv6 loopback address, instead of ipv4:
```
$ quarkus dev
[INFO] Scanning for projects...
[INFO] 
[INFO] -----------------------< <redacted>:<redacted>>------------------------
[INFO] Building <redacted>1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:2.16.4.Final:dev (default-cli) @ <redacted>---
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:2.6:resources @ <redacted>
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] Invoking io.quarkus.platform:quarkus-maven-plugin:2.16.4.Final:generate-code @ <redacted>
[INFO] Invoking org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile @ <redacted>
[INFO] Nothing to compile - all classes are up to date
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:2.6:testResources @ <redacted>
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/imago/Projects/<redacted>/src/test/resources
[INFO] Invoking io.quarkus.platform:quarkus-maven-plugin:2.16.4.Final:generate-code-tests @ <redacted>
[INFO] Invoking org.apache.maven.plugins:maven-compiler-plugin:3.10.1:testCompile @ <redacted>
[INFO] Nothing to compile - all classes are up to date
Listening for transport dt_socket at address: 5005
2023-03-08 06:52:16,892 INFO  [com.git.eir.mav.plu.fro.lib.NodeInstaller] (build-10) Node v16.17.0 is already installed.
2023-03-08 06:52:18,601 INFO  [io.qua.qui.dep.ForwardedDevProcessor] (build-8) Quinoa package manager live coding is up and running on port: 5173 (in 1575ms)
2023-03-08 06:52:18,604 INFO  [io.qua.qui.dep.ForwardedDevProcessor] (build-13) Quinoa is forwarding unhandled requests to port: 5173
__  ____  __  _____   ___  __ ____  ______ 
 --/ __ \/ / / / _ | / _ \/ //_/ / / / __/ 
 -/ /_/ / /_/ / __ |/ , _/ ,< / /_/ /\ \   
--\___\_\____/_/ |_/_/|_/_/|_|\____/___/   
2023-03-08 06:52:19,128 WARN  [io.qua.config] (Quarkus Main Thread) Unrecognized configuration key "quarkus.quinoa.dev-server.host" was provided; it will be ignored; verify that the dependency extension for this configuration is set or that you did not make a typo <- Here I "guessed" how I maybe could target the host
2023-03-08 06:52:19,449 INFO  [io.quarkus] (Quarkus Main Thread) <redacted>1.0.0-SNAPSHOT on JVM (powered by Quarkus 2.16.4.Final) started in 3.298s. Listening on: http://0.0.0.0:8080
2023-03-08 06:52:19,450 INFO  [io.quarkus] (Quarkus Main Thread) Profile dev activated. Live Coding activated.
2023-03-08 06:52:19,450 INFO  [io.quarkus] (Quarkus Main Thread) Installed features: [cdi, quinoa, resteasy-reactive, smallrye-context-propagation, vertx]
2023-03-08 06:52:29,068 ERROR [io.qua.qui.QuinoaDevProxyHandler] (vert.x-eventloop-thread-2) Quinoa failed to forward request, see logs.: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: /[0:0:0:0:0:0:0:1]:5173
Caused by: java.net.ConnectException: Connection refused
        at java.base/sun.nio.ch.Net.pollConnect(Native Method)
        at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
        at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946)
        at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:337)
        at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:334)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:776)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.base/java.lang.Thread.run(Thread.java:833)

```

My vue app is reachable on 127.0.0.1:5173 
```
$ curl "http://localhost:5173"   
<!DOCTYPE html>
<html lang="en">
  <head>
    <script type="module" src="/@vite/client"></script>

    <meta charset="UTF-8">
    <link rel="icon" href="/favicon.ico">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Vite App</title>
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="/src/main.ts"></script>
  </body>
</html
```

But it's not reachable on [::1]:5173
```
$ curl -6 "http://[::1]:5173"
curl: (7) Failed to connect to ::1 port 5173: Connection refused
```

So an alternative is to figure out how to make npm host ipv6 servers, but that's somewhat useless as this npm server only will be used for local debugging. I'd prefer if I could make Quinoa target the ipv4 address, or force it to enter "ipv4-mode" somehow. Is this possible?

Current environment is:
```
$ cat /etc/os-release 
NAME="Ubuntu"
VERSION="20.04 LTS (Focal Fossa)"
...
$ java --version
openjdk 17.0.6 2023-01-17
OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10)
OpenJDK 64-Bit Server VM Temurin-17.0.6+10 (build 17.0.6+10, mixed mode, sharing)
$ quarkus version
2.16.4.Final
```

Many thanks,

Imago

--- a/deployment/src/main/java/io/quarkiverse/quinoa/deployment/DevServerConfig.java
+++ b/deployment/src/main/java/io/quarkiverse/quinoa/deployment/DevServerConfig.java
@@ -34,6 +34,12 @@ public class DevServerConfig {
     public OptionalInt port;
 
     /**
+     * Host of the server to forward requests to.
+     * By default, Quinoa will use the host contained in the request headers.
+     */
+    public Optional<String> host;
+
+    /**
      * After start, Quinoa wait for the external dev server.
      * by sending GET requests to this path waiting for a 200 status.
      *

--- a/deployment/src/main/java/io/quarkiverse/quinoa/deployment/ForwardedDevProcessor.java
+++ b/deployment/src/main/java/io/quarkiverse/quinoa/deployment/ForwardedDevProcessor.java
@@ -73,6 +73,7 @@ public class ForwardedDevProcessor {
         }
         QuinoaConfig oldConfig = liveReload.getContextObject(QuinoaConfig.class);
         liveReload.setContextObject(QuinoaConfig.class, quinoaConfig);
+        final String devServerHost = quinoaConfig.devServer.host.orElse(null);
         if (devService != null) {
             boolean shouldShutdownTheBroker = !quinoaConfig.equals(oldConfig);
             if (!shouldShutdownTheBroker) {
@@ -82,7 +83,7 @@ public class ForwardedDevProcessor {
                 }
                 LOG.debug("Quinoa config did not change; no need to restart.");
                 devServices.produce(devService.toBuildItem());
-                return new ForwardedDevServerBuildItem(quinoaConfig.devServer.port.getAsInt());
+                return new ForwardedDevServerBuildItem(devServerHost, quinoaConfig.devServer.port.getAsInt());
             }
             shutdownDevService();
         }
@@ -109,7 +110,7 @@ public class ForwardedDevProcessor {
         final String checkPath = quinoaConfig.devServer.checkPath.orElse(null);
         if (!quinoaConfig.devServer.managed) {
             if (PackageManager.isDevServerUp(checkPath, devServerPort)) {
-                return new ForwardedDevServerBuildItem(devServerPort);
+                return new ForwardedDevServerBuildItem(devServerHost, devServerPort);
             } else {
                 throw new IllegalStateException(
                         "The Web UI dev server (configured as not managed by Quinoa) is not started on port: " + devServerPort);
@@ -147,7 +148,7 @@ public class ForwardedDevProcessor {
             devService = new DevServicesResultBuildItem.RunningDevService(
                     "quinoa-node-dev-process", null, onClose, Collections.emptyMap());
             devServices.produce(devService.toBuildItem());
-            return new ForwardedDevServerBuildItem(devServerPort);
+            return new ForwardedDevServerBuildItem(devServerHost, devServerPort);
         } catch (Throwable t) {
             packageManager.stopDev(dev.get());
             compressor.closeAndDumpCaptured();
@@ -174,7 +175,8 @@ public class ForwardedDevProcessor {
             LOG.infof("Quinoa is forwarding unhandled requests to port: %d", quinoaConfig.devServer.port.getAsInt());
             final QuinoaHandlerConfig handlerConfig = quinoaConfig.toHandlerConfig(false, httpBuildTimeConfig);
             routes.produce(RouteBuildItem.builder().orderedRoute("/*", QUINOA_ROUTE_ORDER)
-                    .handler(recorder.quinoaProxyDevHandler(handlerConfig, vertx.getVertx(), devProxy.get().getPort(),
+                    .handler(recorder.quinoaProxyDevHandler(handlerConfig, vertx.getVertx(), devProxy.get().getHost(),
+                            devProxy.get().getPort(),
                             quinoaConfig.devServer.websocket))
                     .build());
             if (quinoaConfig.devServer.websocket) {

--- a/deployment/src/main/java/io/quarkiverse/quinoa/deployment/ForwardedDevServerBuildItem.java
+++ b/deployment/src/main/java/io/quarkiverse/quinoa/deployment/ForwardedDevServerBuildItem.java
@@ -4,10 +4,16 @@ import io.quarkus.builder.item.SimpleBuildItem;
 
 public final class ForwardedDevServerBuildItem extends SimpleBuildItem {
 
+    private final String host;
     private final Integer port;
 
-    public ForwardedDevServerBuildItem(Integer port) {
+    public ForwardedDevServerBuildItem(String host, Integer port) {
+        this.host = host;
         this.port = port;
+    }
+
+    public String getHost() {
+        return host;
     }
 
     public Integer getPort() {

--- a/docs/modules/ROOT/pages/includes/attributes.adoc
+++ b/docs/modules/ROOT/pages/includes/attributes.adoc
@@ -1,5 +1,5 @@
-:quarkus-version: 3.0.0.Alpha6
-:quarkus-quinoa-version: 1.2.6
+:quarkus-version: 2.16.4.Final
+:quarkus-quinoa-version: 1.2.7
 :maven-version: 3.8.1+
 
 :quarkus-org-url: https://github.com/quarkusio

--- a/docs/modules/ROOT/pages/includes/quarkus-quinoa.adoc
+++ b/docs/modules/ROOT/pages/includes/quarkus-quinoa.adoc
@@ -410,6 +410,20 @@ endif::add-copy-button-to-env-var[]
 |
 
 
+a|icon:lock[title=Fixed at build time] [[quarkus-quinoa_quarkus.quinoa.dev-server.host]]`link:#quarkus-quinoa_quarkus.quinoa.dev-server.host[quarkus.quinoa.dev-server.host]`
+
+[.description]
+--
+ifdef::add-copy-button-to-env-var[]
+Environment variable: env_var_with_copy_button:+++QUARKUS_QUINOA_DEV_SERVER_HOST+++[]
+endif::add-copy-button-to-env-var[]
+ifndef::add-copy-button-to-env-var[]
+Environment variable: `+++QUARKUS_QUINOA_DEV_SERVER_HOST+++`
+endif::add-copy-button-to-env-var[]
+--|string 
+|
+
+
 a|icon:lock[title=Fixed at build time] [[quarkus-quinoa_quarkus.quinoa.dev-server.check-path]]`link:#quarkus-quinoa_quarkus.quinoa.dev-server.check-path[quarkus.quinoa.dev-server.check-path]`
 
 [.description]

--- a/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java
+++ b/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevProxyHandler.java
@@ -7,6 +7,7 @@ import static io.quarkiverse.quinoa.QuinoaRecorder.resolvePath;
 import static io.quarkiverse.quinoa.QuinoaRecorder.shouldHandleMethod;
 
 import java.util.List;
+import java.util.Optional;
 
 import org.jboss.logging.Logger;
 
@@ -29,16 +30,19 @@ class QuinoaDevProxyHandler implements Handler<RoutingContext> {
             HttpHeaders.CONTENT_RANGE.toString(),
             HttpHeaders.CONTENT_LENGTH.toString(),
             HttpHeaders.CONTENT_TYPE.toString());
+
+    private final Optional<String> host;
     private final int port;
     private final WebClient client;
     private final QuinoaDevWebSocketProxyHandler wsUpgradeHandler;
     private final ClassLoader currentClassLoader;
     private final QuinoaHandlerConfig config;
 
-    QuinoaDevProxyHandler(final QuinoaHandlerConfig config, final Vertx vertx, int port, boolean websocket) {
+    QuinoaDevProxyHandler(final QuinoaHandlerConfig config, final Vertx vertx, String host, int port, boolean websocket) {
+        this.host = Optional.ofNullable(host);
         this.port = port;
         this.client = WebClient.create(vertx);
-        this.wsUpgradeHandler = websocket ? new QuinoaDevWebSocketProxyHandler(vertx, port) : null;
+        this.wsUpgradeHandler = websocket ? new QuinoaDevWebSocketProxyHandler(vertx, host, port) : null;
         this.config = config;
         currentClassLoader = Thread.currentThread().getContextClassLoader();
     }
@@ -88,7 +92,7 @@ class QuinoaDevProxyHandler implements Handler<RoutingContext> {
         headers.remove("Accept");
         // Disable compression in the forwarded request
         headers.remove("Accept-Encoding");
-        client.request(request.method(), port, request.localAddress().host(), uri)
+        client.request(request.method(), port, host.orElseGet(() -> computeHostName(request)), uri)
                 .putHeaders(headers)
                 .send(event -> {
                     if (event.succeeded()) {
@@ -107,6 +111,14 @@ class QuinoaDevProxyHandler implements Handler<RoutingContext> {
                         error(event, ctx);
                     }
                 });
+    }
+
+    static String computeHostName(HttpServerRequest request) {
+        final String[] split = request.host().split(":");
+        if (split.length == 0) {
+            throw new IllegalStateException("Invalid host: " + request.host());
+        }
+        return split[0];
     }
 
     private String computeResourceURI(String path, HttpServerRequest request) {

--- a/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevWebSocketProxyHandler.java
+++ b/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaDevWebSocketProxyHandler.java
@@ -1,7 +1,10 @@
 package io.quarkiverse.quinoa;
 
+import static io.quarkiverse.quinoa.QuinoaDevProxyHandler.computeHostName;
+
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Optional;
 import java.util.concurrent.atomic.AtomicReference;
 
 import org.jboss.logging.Logger;
@@ -18,10 +21,12 @@ import io.vertx.ext.web.RoutingContext;
 class QuinoaDevWebSocketProxyHandler {
     private static final Logger LOG = Logger.getLogger(QuinoaDevWebSocketProxyHandler.class);
     private final HttpClient httpClient;
+    private Optional<String> configuredHost;
     private final int port;
 
-    QuinoaDevWebSocketProxyHandler(Vertx vertx, int port) {
+    QuinoaDevWebSocketProxyHandler(Vertx vertx, String host, int port) {
         this.httpClient = vertx.createHttpClient();
+        this.configuredHost = Optional.ofNullable(host);
         this.port = port;
     }
 
@@ -30,7 +35,7 @@ class QuinoaDevWebSocketProxyHandler {
         ctx.request().pause();
         request.toWebSocket(r -> {
             if (r.succeeded()) {
-                final String host = request.localAddress().host();
+                final String host = configuredHost.orElseGet(() -> computeHostName(request));
                 final String forwardUri = request.uri();
                 LOG.debugf("Quinoa Dev WebSocket Server Connected: %s:%s%s", host, port, forwardUri);
                 final ServerWebSocket serverWs = r.result();

--- a/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaRecorder.java
+++ b/runtime/src/main/java/io/quarkiverse/quinoa/QuinoaRecorder.java
@@ -26,9 +26,9 @@ public class QuinoaRecorder {
     public static final Set<HttpMethod> HANDLED_METHODS = Set.of(HttpMethod.HEAD, HttpMethod.OPTIONS, HttpMethod.GET);
 
     public Handler<RoutingContext> quinoaProxyDevHandler(final QuinoaHandlerConfig handlerConfig, Supplier<Vertx> vertx,
-            int port, boolean websocket) {
+            String host, int port, boolean websocket) {
         logIgnoredPathPrefixes(handlerConfig.ignoredPathPrefixes);
-        return new QuinoaDevProxyHandler(handlerConfig, vertx.get(), port, websocket);
+        return new QuinoaDevProxyHandler(handlerConfig, vertx.get(), host, port, websocket);
     }
 
     public Handler<RoutingContext> quinoaSPARoutingHandler(final QuinoaHandlerConfig handlerConfig) throws IOException {

